src/sinks/util/tcp.rs

```diff
@@ -37,82 +39,182 @@ use tokio_util::codec::{BytesCodec, FramedWrite};
+        source: crate::dns::DnsError,
+    },
+    #[snafu(display("No addresses returned."))]
+    NoAddresses,
+    SendError {
+        source: tokio::io::Error,
+    },
+}
+
 impl TcpSinkConfig {
-    pub fn new(address: String, encoding: EncodingConfig<Encoding>) -> Self {
-        Self {
-            address,
-            encoding,
-            tls: None,
-        }
+    pub fn new(address: String) -> Self {
+        Self { address, tls: None }
     }
 
-    pub fn build(&self, cx: SinkContext) -> crate::Result<(VectorSink, Healthcheck)> {
+    fn build_connector(&self, cx: SinkContext) -> crate::Result<TcpConnector> {
         let uri = self.address.parse::<http::Uri>()?;
 
         let host = uri.host().ok_or(SinkBuildError::MissingHost)?.to_string();
         let port = uri.port_u16().ok_or(SinkBuildError::MissingPort)?;
 
         let tls = MaybeTlsSettings::from_config(&self.tls, false)?;
 
-        let tcp = TcpSink::new(host, port, cx.resolver(), tls);
-        let healthcheck = tcp.healthcheck();
+        let connector = TcpConnector::new(host, port, cx.resolver(), tls);
+
+        Ok(connector)
+    }
+
+    pub fn build_service(&self, cx: SinkContext) -> crate::Result<(TcpService, Healthcheck)> {
+        let connector = self.build_connector(cx)?;
+        let healthcheck = connector.healthcheck();
+        Ok((connector.into(), healthcheck))
+    }
+
+    pub fn build(
+        &self,
+        cx: SinkContext,
+        encoding: EncodingConfig<Encoding>,
+    ) -> crate::Result<(VectorSink, Healthcheck)> {
+        let connector = self.build_connector(cx.clone())?;
+        let healthcheck = connector.healthcheck();
+        let sink: TcpSink = connector.into();
+        let sink = StreamSinkOld::new(sink, cx.acker())
+            .with_flat_map(move |event| iter_ok(encode_event(event, &encoding)));
+
+        Ok((VectorSink::Futures01Sink(Box::new(sink)), healthcheck))
+    }
+}
+
+impl TcpConnector {
+    fn new(host: String, port: u16, resolver: Resolver, tls: MaybeTlsSettings) -> Self {
+        Self {
+            host,
+            port,
+            resolver,
+            tls,
+        }
+    }
+
+    fn connect(&self) -> BoxFuture<'static, Result<TcpOrTlsStream, TcpError>> {
+        let host = self.host.clone();
+        let port = self.port;
+        let resolver = self.resolver;
+        let tls = self.tls.clone();
+
+        async move {
+            let ip = resolver
+                .lookup_ip(host.clone())
+                .await
+                .context(DnsError)?
+                .next()
+                .ok_or(TcpError::NoAddresses)?;
+
+            let addr = SocketAddr::new(ip, port);
+            let stream = tls.connect(host, addr).await.context(ConnectError)?;
+            Ok(FramedWrite::new(stream, BytesCodec::new()))
+        }
+        .boxed()
+    }
+
+    fn healthcheck(&self) -> BoxFuture<'static, crate::Result<()>> {
+        self.connect().map_ok(|_| ()).map_err(|e| e.into()).boxed()
```

Comment:
Can this use `.map_err(Into::into)` too, or is `e` a reference here? I would have expected clippy to catch this.

Reply:
It missed because `redundant_closure_for_method_calls` is in pedantic (moved recently, I guess?).

https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure_for_method_calls